csrc/instance_norm_nvfuser_kernel.cu

```diff
@@ -0,0 +1,269 @@
+      value *= 0x01000193;
+    }
+    return (size_t)value;
+  }
+};
+
+template <typename Params>
+struct ParamsEqual {
+  // Params must be a POD because we read out its memory
+  // contenst as char* when comparing
+  static_assert(std::is_pod<Params>::value, "Params is not POD");
+
+  bool operator()(const Params& a, const Params& b) const {
+    auto ptr1 = reinterpret_cast<const uint8_t*>(&a);
+    auto ptr2 = reinterpret_cast<const uint8_t*>(&b);
+    return memcmp(ptr1, ptr2, sizeof(Params)) == 0;
+  }
+};
+
+using namespace torch::jit::fuser::cuda;
+using namespace at::indexing;
+
+// Make a tensor that is known to be fully contiguous of dimensionality=ndims,
+// but unknown sizes
+TensorView* makeContigTensor(size_t ndims, DataType dtype = DataType::Float) {
+  return TensorViewBuilder()
+      .ndims(ndims)
+      .dtype(dtype)
+      .contiguity(std::vector<bool>(ndims, true))
+      .build();
+}
+
+struct InstanceNormKey {
+  c10::ScalarType input_dtype;//int8_t dtype;
+  c10::ScalarType weight_dtype;
+  c10::ScalarType mean_dtype;
+  size_t dim;
+  bool channels_last;
+  bool running_mean;
+  bool affine;
+  float eps;
+};
+
+auto get_dtype(c10::ScalarType dtype) {
+  auto ret_dtype = DataType::Float;
+  if (dtype == c10::ScalarType::Double) {
+    ret_dtype = DataType::Double;
+  } else if (dtype == c10::ScalarType::Half) {
+    ret_dtype = DataType::Half;
+  } else if (dtype == c10::ScalarType::BFloat16) {
+    ret_dtype = DataType::BFloat16;
+  }
+  return ret_dtype;
+}
+
+// TODO: doesn't support all combinations of dtype e.g., bias, run_var, ..
+// bias is assumed to match weight, run_var is assumed to match run_mean
+void getKey(const at::Tensor& input, const at::Tensor& weight, const at::Tensor& run_mean, const bool channels_last, const float eps, InstanceNormKey& key) {
+  memset(&key, 0, sizeof(InstanceNormKey));
+  key.input_dtype = input.scalar_type();// static_cast<int8_t>(input.scalar_type());
+  key.weight_dtype = weight.scalar_type();
+  key.mean_dtype = run_mean.scalar_type();
+  key.dim = input.sizes().size();
+  key.channels_last = channels_last;
+  key.eps = eps;
+  key.running_mean = run_mean.sizes().size() > 0;
+  key.affine = weight.sizes().size() ? true : false;
+}
+
+std::unordered_map<InstanceNormKey, std::unique_ptr<FusionExecutorCache>, ParamsHash<InstanceNormKey>, ParamsEqual<InstanceNormKey> > forward_fusion_cache;
+std::unordered_map<InstanceNormKey, std::unique_ptr<FusionExecutorCache>, ParamsHash<InstanceNormKey>, ParamsEqual<InstanceNormKey> > backward_fusion_cache;
+
+std::vector<at::Tensor> instance_norm_nvfuser_forward(
+  at::Tensor input,
+  at::Tensor weight,
+  at::Tensor bias,
+  at::Tensor run_mean,
+  at::Tensor run_var,
+  const bool use_input_stats,
+  const float momentum,
+  const float eps,
+  const bool channels_last) {
+    InstanceNormKey forward_key;
+    memset(&forward_key, 0, sizeof(InstanceNormKey));
+    getKey(input, weight, run_mean, channels_last, eps, forward_key);
```

Comment:
why is `eps` part of the key, but `momentum` isn't? Also, can they be passed as args and not constants, so that the kernel doesn't have to be recompiled?

Reply:
Neither should need to be part of the key.